File: js/src/utils/nodeBuilder.ts
/**
 * NodeBuilder utilities for common node configuration patterns.
 * 
 * These utilities provide convenient ways to create pre-configured node data
 * without requiring custom JavaScript code. All configurations can be used
 * directly from Python by constructing the appropriate data structures.
 */

import type { 
  CustomNodeData, 
  NodeHeaderConfig, 
  NodeFooterConfig, 
  NodeStyleConfig,
  FieldConfig,
  ValidationConfig 
} from "../types/schema";

/**
 * Builder class for creating configured nodes with a fluent API
 */
export class NodeBuilder {
  private data: Partial<CustomNodeData>;

  constructor(label: string) {
    this.data = {
      label,
      values: {},
    };
  }

  /**
   * Set the layout type
   */
  withLayout(layoutType: string): this {
    this.data.layoutType = layoutType;
    return this;
  }

  /**
   * Set handle types
   */
  withHandles(
    type: "base" | "button" | "labeled",
    inputType?: "base" | "button" | "labeled",
    outputType?: "base" | "button" | "labeled"
  ): this {
    this.data.handleType = type;
    if (inputType) this.data.inputHandleType = inputType;
    if (outputType) this.data.outputHandleType = outputType;
    return this;
  }

  /**
   * Configure the header
   */
  withHeader(config: NodeHeaderConfig): this {
    this.data.header = config;
    return this;
  }

  /**
   * Configure the footer
   */
  withFooter(config: NodeFooterConfig): this {
    this.data.footer = config;
    return this;
  }

  /**
   * Configure styles
   */
  withStyle(config: NodeStyleConfig): this {
    this.data.style = config;
    return this;
  }

  /**
   * Configure validation
   */
  withValidation(config: ValidationConfig): this {
    this.data.validation = config;
    return this;
  }

  /**
   * Set field configurations
   */
  withFieldConfigs(configs: Record<string, FieldConfig>): this {
    this.data.fieldConfigs = configs;
    return this;
  }

  /**
   * Add a single field configuration
   */
  withFieldConfig(fieldKey: string, config: FieldConfig): this {
    if (!this.data.fieldConfigs) {
      this.data.fieldConfigs = {};
    }
    this.data.fieldConfigs[fieldKey] = config;
    return this;
  }

  /**
   * Build the final node data
   */
  build(): Partial<CustomNodeData> {
    return this.data;
  }
}

/**
 * Pre-configured node templates
 */

/**
 * Create a minimal node configuration (no header, compact layout)
 */
export function createMinimalNode(label: string): Partial<CustomNodeData> {
  return {
    label,
    header: { show: false },
    style: {
      minWidth: "150px",
      shadow: "sm",
      className: "border",
    },
    layoutType: "compact",
    values: {},
  };
}

/**
 * Create a debug/inspector node (shows all internal state)
 */
export function createDebugNode(label: string = "Debug"): Partial<CustomNodeData> {
  return {
    label,
    header: {
      show: true,
      icon: "🔍",
      className: "bg-yellow-500 text-yellow-950",
    },
    footer: {
      show: true,
      text: "Debug mode enabled",
      className: "bg-yellow-50 text-yellow-800 border-yellow-200",
    },
    style: {
      className: "border-yellow-400",
      shadow: "lg",
    },
    validation: {
      showErrors: true,
      errorPosition: "inline",
      validateOnChange: true,
    },
    values: {},
  };
}

/**
 * Create a form-heavy node (optimized for many input fields)
 */
export function createFormNode(label: string): Partial<CustomNodeData> {
  return {
    label,
    header: {
      show: true,
      className: "bg-blue-600 text-white",
    },
    layoutType: "vertical",
    style: {
      minWidth: "300px",
      maxWidth: "500px",
      shadow: "md",
    },
    validation: {
      showErrors: true,
      errorPosition: "inline",
    },
    values: {},
  };
}

/**
 * Create a compact processing node (minimal UI, focus on I/O)
 */
export function createProcessingNode(label: string, icon?: string): Partial<CustomNodeData> {
  return {
    label,
    header: {
      show: true,
      icon: icon || "⚙️",
      className: "bg-gray-700 text-white",
    },
    layoutType: "horizontal",
    handleType: "button",
    style: {
      minWidth: "180px",
      shadow: "md",
    },
    values: {},
  };
}

/**
 * Create a data source node (emphasis on outputs)
 */
export function createSourceNode(label: string, icon?: string): Partial<CustomNodeData> {
  return {
    label,
    header: {
      show: true,
      icon: icon || "📥",
      className: "bg-green-600 text-white",
    },
    layoutType: "vertical",
    handleType: "labeled",
    style: {
      minWidth: "200px",
      shadow: "md",
      className: "border-green-500",
    },
    values: {},
  };
}

/**
 * Create a data sink node (emphasis on inputs)
 */
export function createSinkNode(label: string, icon?: string): Partial<CustomNodeData> {
  return {
    label,
    header: {
      show: true,
      icon: icon || "📤",
      className: "bg-red-600 text-white",
    },
    layoutType: "vertical",
    handleType: "labeled",
    style: {
      minWidth: "200px",
      shadow: "md",
      className: "border-red-500",
    },
    values: {},
  };
}

/**
 * Create a visualization node (larger, emphasis on display)
 */
export function createVisualizationNode(label: string): Partial<CustomNodeData> {
  return {
    label,
    header: {
      show: true,
      icon: "📊",
      className: "bg-purple-600 text-white",
    },
    layoutType: "vertical",
    style: {
      minWidth: "400px",
      shadow: "lg",
    },
    values: {},
  };
}

/**
 * Configuration preset for conditional field visibility
 */
export function createConditionalFields(
  triggerField: string,
  triggerValue: any,
  dependentFields: string[]
): Record<string, FieldConfig> {
  const configs: Record<string, FieldConfig> = {};
  
  for (const field of dependentFields) {
    configs[field] = {
      showWhen: {
        field: triggerField,
        operator: "equals",
        value: triggerValue,
      },
    };
  }
  
  return configs;
}

/**
 * Create a preset for readonly/display-only nodes
 */
export function createReadonlyNode(label: string): Partial<CustomNodeData> {
  return {
    label,
    header: {
      show: true,
      icon: "📋",
      className: "bg-slate-500 text-white",
    },
    style: {
      className: "opacity-90 border-slate-300",
      shadow: "sm",
    },
    values: {},
  };
}

/**
 * Apply readonly configuration to all fields
 */
export function makeAllFieldsReadonly(
  fieldKeys: string[]
): Record<string, FieldConfig> {
  const configs: Record<string, FieldConfig> = {};
  
  for (const key of fieldKeys) {
    configs[key] = { readonly: true };
  }
  
  return configs;
}

/**
 * Helper to merge node configurations
 */
export function mergeNodeConfigs(
  base: Partial<CustomNodeData>,
  override: Partial<CustomNodeData>
): Partial<CustomNodeData> {
  return {
    ...base,
    ...override,
    header: { ...base.header, ...override.header },
    footer: { ...base.footer, ...override.footer },
    style: { ...base.style, ...override.style },
    validation: { ...base.validation, ...override.validation },
    fieldConfigs: { ...base.fieldConfigs, ...override.fieldConfigs },
    values: { ...base.values, ...override.values },
  };
}

File: src/pynodewidget/node_builder.py
"""Node builder utilities for creating configured nodes from Python.

This module provides helper functions to create commonly-used node configurations
without writing custom JavaScript. All configurations are pure Python dictionaries
that get serialized and passed to the JsonSchemaNode component.

Example:
    >>> from pynodewidget import JsonSchemaNodeWidget
    >>> from pynodewidget.node_builder import create_minimal_node, create_form_node
    >>> 
    >>> # Create a minimal node
    >>> config = create_minimal_node("Process")
    >>> node = JsonSchemaNodeWidget.from_pydantic(
    ...     MyModel,
    ...     **config
    ... )
"""

from typing import Dict, Any, List, Optional, Literal


# Type aliases for clarity
FieldValue = str | int | float | bool | None
HandleType = Literal["base", "button", "labeled"]
ShadowSize = Literal["sm", "md", "lg", "xl", "none"]
ErrorPosition = Literal["inline", "tooltip", "footer"]
ConditionOperator = Literal["equals", "notEquals", "greaterThan", "lessThan", "contains"]


def create_minimal_node(
    label: str,
    layout_type: str = "compact",
    min_width: str = "150px"
) -> Dict[str, Any]:
    """Create a minimal node configuration (no header, compact layout).
    
    Args:
        label: Node display label
        layout_type: Layout to use (default: "compact")
        min_width: Minimum node width (default: "150px")
        
    Returns:
        Configuration dictionary for JsonSchemaNodeWidget
    """
    return {
        "label": label,
        "layout_type": layout_type,
        "header": {"show": False},
        "style": {
            "minWidth": min_width,
            "shadow": "sm",
            "className": "border",
        },
    }


def create_debug_node(label: str = "Debug") -> Dict[str, Any]:
    """Create a debug/inspector node (shows all internal state).
    
    Args:
        label: Node display label (default: "Debug")
        
    Returns:
        Configuration dictionary for JsonSchemaNodeWidget
    """
    return {
        "label": label,
        "header": {
            "show": True,
            "icon": "🔍",
            "className": "bg-yellow-500 text-yellow-950",
        },
        "footer": {
            "show": True,
            "text": "Debug mode enabled",
            "className": "bg-yellow-50 text-yellow-800 border-yellow-200",
        },
        "style": {
            "className": "border-yellow-400",
            "shadow": "lg",
        },
        "validation": {
            "showErrors": True,
            "errorPosition": "inline",
            "validateOnChange": True,
        },
    }


def create_form_node(
    label: str,
    min_width: str = "300px",
    max_width: str = "500px"
) -> Dict[str, Any]:
    """Create a form-heavy node (optimized for many input fields).
    
    Args:
        label: Node display label
        min_width: Minimum node width (default: "300px")
        max_width: Maximum node width (default: "500px")
        
    Returns:
        Configuration dictionary for JsonSchemaNodeWidget
    """
    return {
        "label": label,
        "layout_type": "vertical",
        "header": {
            "show": True,
            "className": "bg-blue-600 text-white",
        },
        "style": {
            "minWidth": min_width,
            "maxWidth": max_width,
            "shadow": "md",
        },
        "validation": {
            "showErrors": True,
            "errorPosition": "inline",
        },
    }


def create_processing_node(
    label: str,
    icon: str = "⚙️",
    handle_type: HandleType = "button"
) -> Dict[str, Any]:
    """Create a compact processing node (minimal UI, focus on I/O).
    
    Args:
        label: Node display label
        icon: Icon emoji (default: "⚙️")
        handle_type: Type of handles to use (default: "button")
        
    Returns:
        Configuration dictionary for JsonSchemaNodeWidget
    """
    return {
        "label": label,
        "layout_type": "horizontal",
        "handle_type": handle_type,
        "header": {
            "show": True,
            "icon": icon,
            "className": "bg-gray-700 text-white",
        },
        "style": {
            "minWidth": "180px",
            "shadow": "md",
        },
    }


def create_source_node(
    label: str,
    icon: str = "📥",
    handle_type: HandleType = "labeled"
) -> Dict[str, Any]:
    """Create a data source node (emphasis on outputs).
    
    Args:
        label: Node display label
        icon: Icon emoji (default: "📥")
        handle_type: Type of handles to use (default: "labeled")
        
    Returns:
        Configuration dictionary for JsonSchemaNodeWidget
    """
    return {
        "label": label,
        "layout_type": "vertical",
        "handle_type": handle_type,
        "header": {
            "show": True,
            "icon": icon,
            "className": "bg-green-600 text-white",
        },
        "style": {
            "minWidth": "200px",
            "shadow": "md",
            "className": "border-green-500",
        },
    }


def create_sink_node(
    label: str,
    icon: str = "📤",
    handle_type: HandleType = "labeled"
) -> Dict[str, Any]:
    """Create a data sink node (emphasis on inputs).
    
    Args:
        label: Node display label
        icon: Icon emoji (default: "📤")
        handle_type: Type of handles to use (default: "labeled")
        
    Returns:
        Configuration dictionary for JsonSchemaNodeWidget
    """
    return {
        "label": label,
        "layout_type": "vertical",
        "handle_type": handle_type,
        "header": {
            "show": True,
            "icon": icon,
            "className": "bg-red-600 text-white",
        },
        "style": {
            "minWidth": "200px",
            "shadow": "md",
            "className": "border-red-500",
        },
    }


def create_visualization_node(
    label: str,
    min_width: str = "400px"
) -> Dict[str, Any]:
    """Create a visualization node (larger, emphasis on display).
    
    Args:
        label: Node display label
        min_width: Minimum node width (default: "400px")
        
    Returns:
        Configuration dictionary for JsonSchemaNodeWidget
    """
    return {
        "label": label,
        "layout_type": "vertical",
        "header": {
            "show": True,
            "icon": "📊",
            "className": "bg-purple-600 text-white",
        },
        "style": {
            "minWidth": min_width,
            "shadow": "lg",
        },
    }


def create_readonly_node(label: str) -> Dict[str, Any]:
    """Create a preset for readonly/display-only nodes.
    
    Args:
        label: Node display label
        
    Returns:
        Configuration dictionary for JsonSchemaNodeWidget
    """
    return {
        "label": label,
        "header": {
            "show": True,
            "icon": "📋",
            "className": "bg-slate-500 text-white",
        },
        "style": {
            "className": "opacity-90 border-slate-300",
            "shadow": "sm",
        },
    }


def with_custom_header(
    config: Dict[str, Any],
    icon: Optional[str] = None,
    bg_color: Optional[str] = None,
    text_color: Optional[str] = None,
    class_name: Optional[str] = None,
    show: bool = True
) -> Dict[str, Any]:
    """Add or modify header configuration.
    
    Args:
        config: Base configuration dictionary
        icon: Icon emoji to display
        bg_color: Background color (CSS color string)
        text_color: Text color (CSS color string)
        class_name: Custom CSS classes
        show: Whether to show the header
        
    Returns:
        Updated configuration dictionary
    """
    header_config = config.get("header", {})
    
    if icon is not None:
        header_config["icon"] = icon
    if bg_color is not None:
        header_config["bgColor"] = bg_color
    if text_color is not None:
        header_config["textColor"] = text_color
    if class_name is not None:
        header_config["className"] = class_name
    header_config["show"] = show
    
    return {**config, "header": header_config}


def with_footer(
    config: Dict[str, Any],
    text: str,
    class_name: Optional[str] = None
) -> Dict[str, Any]:
    """Add footer configuration.
    
    Args:
        config: Base configuration dictionary
        text: Footer text to display
        class_name: Custom CSS classes
        
    Returns:
        Updated configuration dictionary
    """
    footer_config = {
        "show": True,
        "text": text,
    }
    if class_name:
        footer_config["className"] = class_name
    
    return {**config, "footer": footer_config}


def with_style(
    config: Dict[str, Any],
    min_width: Optional[str] = None,
    max_width: Optional[str] = None,
    border_radius: Optional[str] = None,
    shadow: Optional[ShadowSize] = None,
    class_name: Optional[str] = None
) -> Dict[str, Any]:
    """Add or modify style configuration.
    
    Args:
        config: Base configuration dictionary
        min_width: Minimum width (e.g., "200px")
        max_width: Maximum width (e.g., "500px")
        border_radius: Border radius (e.g., "8px")
        shadow: Shadow size ("sm", "md", "lg", "xl", "none")
        class_name: Custom CSS classes
        
    Returns:
        Updated configuration dictionary
    """
    style_config = config.get("style", {})
    
    if min_width is not None:
        style_config["minWidth"] = min_width
    if max_width is not None:
        style_config["maxWidth"] = max_width
    if border_radius is not None:
        style_config["borderRadius"] = border_radius
    if shadow is not None:
        style_config["shadow"] = shadow
    if class_name is not None:
        style_config["className"] = class_name
    
    return {**config, "style": style_config}


def with_validation(
    config: Dict[str, Any],
    show_errors: bool = True,
    error_position: ErrorPosition = "inline",
    validate_on_change: bool = False
) -> Dict[str, Any]:
    """Add validation configuration.
    
    Args:
        config: Base configuration dictionary
        show_errors: Whether to show validation errors
        error_position: Where to display errors ("inline", "tooltip", "footer")
        validate_on_change: Validate as user types
        
    Returns:
        Updated configuration dictionary
    """
    return {
        **config,
        "validation": {
            "showErrors": show_errors,
            "errorPosition": error_position,
            "validateOnChange": validate_on_change,
        }
    }


def create_conditional_field(
    trigger_field: str,
    trigger_value: FieldValue,
    operator: ConditionOperator = "equals"
) -> Dict[str, Any]:
    """Create a field condition for conditional visibility.
    
    Args:
        trigger_field: Field name to watch
        trigger_value: Value to compare against
        operator: Comparison operator
        
    Returns:
        Field configuration with showWhen condition
    """
    return {
        "showWhen": {
            "field": trigger_field,
            "operator": operator,
            "value": trigger_value,
        }
    }


def make_field_readonly(field_key: str) -> Dict[str, Dict[str, Any]]:
    """Create readonly configuration for a field.
    
    Args:
        field_key: Field name to make readonly
        
    Returns:
        Field configurations dictionary
    """
    return {field_key: {"readonly": True}}


def make_field_hidden(field_key: str) -> Dict[str, Dict[str, Any]]:
    """Create hidden configuration for a field.
    
    Args:
        field_key: Field name to hide
        
    Returns:
        Field configurations dictionary
    """
    return {field_key: {"hidden": True}}


def make_fields_conditional(
    trigger_field: str,
    trigger_value: FieldValue,
    dependent_fields: List[str],
    operator: ConditionOperator = "equals"
) -> Dict[str, Dict[str, Any]]:
    """Create conditional visibility for multiple fields.
    
    All dependent fields will only show when the trigger field has the specified value.
    
    Args:
        trigger_field: Field name to watch
        trigger_value: Value that triggers visibility
        dependent_fields: List of field names to show conditionally
        operator: Comparison operator (default: "equals")
        
    Returns:
        Field configurations dictionary
    """
    configs = {}
    condition = create_conditional_field(trigger_field, trigger_value, operator)
    
    for field in dependent_fields:
        configs[field] = condition
    
    return configs


def merge_configs(*configs: Dict[str, Any]) -> Dict[str, Any]:
    """Merge multiple configuration dictionaries.
    
    Later configurations override earlier ones. Nested dictionaries like
    header, footer, style are merged recursively.
    
    Args:
        *configs: Configuration dictionaries to merge
        
    Returns:
        Merged configuration dictionary
    """
    result: dict[str, Any] = {}
    
    for config in configs:
        for key, value in config.items():
            if key in ("header", "footer", "style", "validation", "fieldConfigs"):
                # Merge nested dicts
                result[key] = {**result.get(key, {}), **value}
            else:
                result[key] = value
    
    return result
